src/audiobuffer/audiobuffer.cpp

```diff
@@ -0,0 +1,252 @@
+	// //////////////////////////////////////////////////////////////
+	// //////////////////////////////////////////////////////////////
+	// //////////////////////////////////////////////////////////////
+
+	BufferStream::BufferStream() {}
+
+	BufferStream::~BufferStream()
+	{
+		stop();
+	}
+
+	void BufferStream::setBufferStream(
+		Player *aPlayer,
+		ActiveSound *aParent,
+		unsigned int maxBufferSize,
+    	SoLoud::time bufferingTimeNeeds,
+		PCMformat pcmFormat,
+		dartOnBufferingCallback_t onBufferingCallback)
+	{
+		mSampleCount = 0;
+		dataIsEnded = false;
+		mEndianness = Endianness::BUFFER_LITTLE_ENDIAN; // TODO?
+		mThePlayer = aPlayer;
+		mParent = aParent;
+		mPCMformat.sampleRate = pcmFormat.sampleRate;
+		mPCMformat.channels = pcmFormat.channels;
+		mPCMformat.bytesPerSample = pcmFormat.bytesPerSample;
+		mPCMformat.dataType = pcmFormat.dataType;
+		mBuffer.clear();
+		mBuffer.setSizeInBytes(maxBufferSize);
+		mBufferingTimeNeeds = bufferingTimeNeeds;
+		mChannels = pcmFormat.channels;
+		mBaseSamplerate = (float)pcmFormat.sampleRate;
+		mOnBufferingCallback = onBufferingCallback;
+	}
+
+	void BufferStream::setDataIsEnded()
+	{
+		dataIsEnded = true;
+	}
+
+	PlayerErrors BufferStream::addData(const void *aData, unsigned int aDataLen)
+	{
+		if (dataIsEnded)
+		{
+			return pcmBufferFullOrStreamEnded;
+		}
+
+		unsigned int bytesWritten = 0;
+		// add PCM data to the buffer
+		switch (mPCMformat.dataType)
+		{
+		case 0:
+			bytesWritten = mBuffer.addData((float *)aData, aDataLen / mPCMformat.bytesPerSample);
+			break;
+		case 1:
+			bytesWritten = mBuffer.addData((int8_t *)aData, aDataLen / mPCMformat.bytesPerSample);
+			break;
+		case 2:
+			bytesWritten = mBuffer.addData((int16_t *)aData, aDataLen / mPCMformat.bytesPerSample);
+			break;
+		case 3:
+			bytesWritten = mBuffer.addData((int32_t *)aData, aDataLen / mPCMformat.bytesPerSample);
+			break;
+		}
+
+		mSampleCount += bytesWritten / mPCMformat.bytesPerSample;
+
+		// If a handle reaches the end and data is not ended, we have to wait for it has enough data
+		// to reach [TIME_FOR_BUFFERING] and restart playing it.
+		// time currBufferTime = 1.0f * mBuffer.getFloatsBufferSize() / mPCMformat.channels / mPCMformat.sampleRate;
+		time currBufferTime = getLength();
+		for (int i = 0; i < mParent->handle.size(); i++)
+		{
+			double pos = mThePlayer->getPosition(mParent->handle[i].handle);
+			// This handle needs to wait for [TIME_FOR_BUFFERING]
+			if (pos >= currBufferTime && !mThePlayer->getPause(mParent->handle[i].handle))
+			{
+				mParent->handle[i].bufferingTime = currBufferTime;
+				mThePlayer->setPause(mParent->handle[i].handle, true);
+				if (mOnBufferingCallback != nullptr)
+				{
+#ifdef __EMSCRIPTEN__
+						// Call the Dart callback stored on globalThis, if it exists.
+						EM_ASM({
+							// Compose the function name for this soundHash
+							var functionName = "dartOnBufferingCallback_" + $3;
```

Comment:
Oh, here it is. It would be good to at least document where this is coming from, on both sides. Also add something like "If you change this here, you must also change the corresponding string on the {Dart/C++} side: path/to/file."

Reply:
comments added on both sides.
Done.